File: lib/analytics/abc.ts
/**
 * ABC анализ — классификация позиций по их вкладу в кумулятивную метрику.
 * Метрикой обычно служит выручка или маржа, но можно подать любую (профит, количество).
 *
 * Класс A — позиции, дающие первые ~80% (по умолчанию).
 * Класс B — следующие ~15%.
 * Класс C — оставшиеся ~5%.
 */

export type AbcInput = {
  id: string;
  name: string;
  value: number; // метрика (выручка / маржа / штук)
};

export type AbcClass = 'A' | 'B' | 'C';

export type AbcRow = AbcInput & {
  share: number;       // доля от общей метрики
  cumShare: number;    // кумулятивная доля
  rank: number;
  class: AbcClass;
};

export type AbcOptions = {
  thresholds?: { a: number; b: number }; // по умолчанию a=0.8, b=0.95
};

export function buildAbcReport(inputs: AbcInput[], opts: AbcOptions = {}): AbcRow[] {
  const aT = opts.thresholds?.a ?? 0.8;
  const bT = opts.thresholds?.b ?? 0.95;

  const sorted = [...inputs]
    .filter((x) => Number.isFinite(x.value))
    .sort((a, b) => b.value - a.value);

  const total = sorted.reduce((s, x) => s + Math.max(0, x.value), 0);
  if (total <= 0) {
    return sorted.map((r, i) => ({
      ...r,
      share: 0,
      cumShare: 0,
      rank: i + 1,
      class: 'C' as AbcClass,
    }));
  }

  let cum = 0;
  return sorted.map((r, i) => {
    const share = Math.max(0, r.value) / total;
    cum += share;
    const cls: AbcClass = cum <= aT ? 'A' : cum <= bT ? 'B' : 'C';
    return {
      ...r,
      share,
      cumShare: cum,
      rank: i + 1,
      class: cls,
    };
  });
}

export function summarizeAbc(rows: AbcRow[]) {
  const groups: Record<AbcClass, { count: number; value: number; share: number }> = {
    A: { count: 0, value: 0, share: 0 },
    B: { count: 0, value: 0, share: 0 },
    C: { count: 0, value: 0, share: 0 },
  };
  for (const r of rows) {
    groups[r.class].count += 1;
    groups[r.class].value += r.value;
    groups[r.class].share += r.share;
  }
  return groups;
}

File: backend/app/analytics/abc.py
"""
ABC анализ — порт widget/lib/analytics/abc.ts.

Классификация позиций по их вкладу в кумулятивную метрику (по умолчанию —
выручка). Класс A — первые ~80% (порог 0.80), B — до ~95% (порог 0.95),
C — остальное.

Сохраняем те же пороги по умолчанию и ту же логику классификации:
  cls = 'A' if cum <= 0.80 else 'B' if cum <= 0.95 else 'C'

Это специфика TS-реализации: cum обновляется ДО проверки, и первая
позиция уже может быть >0.80 — тогда A окажется пустым. Это
консистентно с тем что показывает виджет.
"""
from __future__ import annotations

import math
from dataclasses import dataclass
from typing import Literal


AbcClass = Literal["A", "B", "C"]


@dataclass(frozen=True)
class AbcInput:
    id: str
    name: str
    value: float


@dataclass
class AbcRow:
    id: str
    name: str
    value: float
    share: float          # доля от total
    cum_share: float      # кумулятивная доля (растёт по списку)
    rank: int             # 1..N
    class_: AbcClass


@dataclass
class AbcThresholds:
    a: float = 0.80
    b: float = 0.95


def build_abc_report(
    inputs: list[AbcInput],
    thresholds: AbcThresholds | None = None,
) -> list[AbcRow]:
    th = thresholds or AbcThresholds()

    sorted_inputs = sorted(
        (x for x in inputs if math.isfinite(x.value)),
        key=lambda r: r.value,
        reverse=True,
    )

    total = sum(max(0.0, x.value) for x in sorted_inputs)
    if total <= 0:
        return [
            AbcRow(
                id=r.id, name=r.name, value=r.value,
                share=0.0, cum_share=0.0, rank=i + 1, class_="C",
            )
            for i, r in enumerate(sorted_inputs)
        ]

    rows: list[AbcRow] = []
    cum = 0.0
    for i, r in enumerate(sorted_inputs):
        share = max(0.0, r.value) / total
        cum += share
        cls: AbcClass = "A" if cum <= th.a else ("B" if cum <= th.b else "C")
        rows.append(
            AbcRow(
                id=r.id, name=r.name, value=r.value,
                share=share, cum_share=cum, rank=i + 1, class_=cls,
            )
        )
    return rows


def summarize_abc(rows: list[AbcRow]) -> dict[str, dict[str, float]]:
    """Эквивалент summarizeAbc из TS — {A: {count, value, share}, ...}."""
    groups = {
        "A": {"count": 0, "value": 0.0, "share": 0.0},
        "B": {"count": 0, "value": 0.0, "share": 0.0},
        "C": {"count": 0, "value": 0.0, "share": 0.0},
    }
    for r in rows:
        g = groups[r.class_]
        g["count"] += 1
        g["value"] += r.value
        g["share"] += r.share
    return groups
